### PR TITLE
KAFKA-7572: Producer should not send requests with negative partition id

### PR DESCRIPTION
**Pull request by ghost**
Partition id should never be a negative value.
This commit will make debug easier, when custom Partitioner generate an invalid negative partition id.

### CODE REVIEW

**Comment by ghost** [comment]
test this please

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -1220,10 +1220,14 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
...
-                partition :
-                partitioner.partition(
-                        record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        if (partition == null) {
+            partition = partitioner.partition(record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        }
+        if (partition >= 0) {
+            return partition;
+        } else {
+            throw new IllegalStateException("Partition id is less than 0, partition_id=" + partition);
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I think it should be a KafkaException (or a subclass of it). IllegalStateException in this class refers to an invalid configuration that is more compile time than runtime, whereas a partitioning problem is purely runtime and only occurs with custom partitioners.

Comment (ghost) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: I followed this error message type, to make it consistent https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java#L75

**Comment by ghost** [comment]
test this please

**Comment by predatorray** [comment]
Hi @viktorsomogyi. Any update on this pull request and the issue KAFKA-7572? It looks like the author's account has been deleted. Can i submit a new pull request on this as i think the code change will be quite simple?

**Comment by viktorsomogyi** [comment]
@predatorray yes, feel free to submit a new PR.

**Comment by predatorray** [comment]
@viktorsomogyi I have opened the PR #10525.

**Pull request closed without merging**